Simkin biomass and productivity function updates

samsimkin opened the pull request.

samsimkin committed:
change biomass and productivity function names and add data tables

samsimkin committed:
some changes to calculateBiomass function

samsimkin committed:
biomass and productivity function name changes

samsimkin committed:
small change

samsimkin committed:
biomass and productivity function updates

samsimkin committed:
remove a few commented out lines

samsimkin committed:
fix site summaries by summing taxa within plots first

samsimkin committed:
exception handling for a line

samsimkin committed:
biomass and productivity function improvements

samsimkin committed:
testthat tests now utilize single list input object but still yield errors with devtools check

samsimkin committed:
reduce size of static ancillary datasets and finished troubleshooting plots with missing metadata

samsimkin committed:
Merge branch 'dev' into biomass

samsimkin committed:
biomass and productivity functions have testthat tests and no issues in devtools check

samsimkin committed:
Merge branch 'biomass' of https://github.com/samsimkin/neonPlants into biomass

samsimkin committed:
split out herb and wood functions and made umbrella mass and prod functions

samsimkin committed:
Merge branch 'dev' of https://github.com/NEONScience/neonPlants into biomass

samsimkin committed:
add purrr for estimateProd and estimateMass functions

samsimkin committed:
Merge pull request #15 from samsimkin/biomass

Simkin biomass and productivity functions

samsimkin committed:
change mass functions to accept df inputs and enable inclusion of roots in umbrella estimateMass function

samsimkin committed:
deal with ghost code

samsimkin committed:
fixed all devtools_check errors

schafnoir commented:
Hi @samsimkin - I added a `year` column to the 'site' output table for the scaleRootMass function. Can you run the scaleRootMass function on the test data in the 'testdata' folder and let me know if the output is correct - i.e., it will work with your umbrella function?

Also, one more request - I see above that there is now a conflict with the DESCRIPTION file, likely because it was modified when I approved Katie's PR. To resolve, can you merge the `dev` branch into your `dev_SMS` branch? 

Finally, I see you've got some tests developed, which is excellent. Have you run `devtools::check()` and all of these tests pass? And there are no other errors identified?

samsimkin committed:
resolve DESCRIPTION file conflict

samsimkin committed:
a few tweaks to accommodate rootMass in umbrella mass function

samsimkin commented:
@schafnoir The estimateMass umbrella function handles the scaleRootMass site output table. Once merged by siteID and year with data from other data products the eventID, startDate, and endDate fields are misleading since they only apply to roots, so for purposes of the umbrella function I've removed them. The rootMass fields with gm2 or gm3 aren't readily combined with data from other data products, so I could also hardcode the umbrella function further to retain just domainID, siteID, year, rootMassMean_Mgha, rootMassSD_Mgha, and rootPlotNum from the root site table, although that would break if you changed the names of any of those fields. What do you think?

I resolved the merge conflict regarding the DESCRIPTION file.

When I run devtools::check() my tests all pass and the only remaining note is with regard to the stackPlantPresence global binding variables.

schafnoir committed:
Review of scaleHerbMass function and code standardization with other functions.

schafnoir committed:
Minor tweaks to accommodate renaming of input arguments in scaleHerbMass function to be consistent with root functions.

schafnoir commented:
Hi @samsimkin  - I think adding further hard-coding into the umbrella function to only select the columns that make sense from the site-level root mass output makes sense. Go for it!

I have also completed review of the `scaleHerbMass` function. I re-worked the roxygen header info for this function to be as consistent as possible with the root and PHE functions, and I standardized the code commenting and formatting. Here is line-by-line feedback:

- [ ] line 161: Average qaDryMass == "Y" values where they exist rather than filtering out. We have no reason to believe one is better than the other. Concatenate any remarks. Use approach like in standardizeRootMass().
- [ ] line 170: The `dplyr::group_by_at()` function appears to be superceded. Replace with just `dplyr::group_by()`? Not sure what problem the `_at` version of the function is trying to solve...
- [ ] line 249: Rather than using a  value of "dryMass_gm2_" for the 'names_prefix' argument, use the 'names_glue' argument and add "_gm2" as a suffix to the column name. The column naming convention would then be consistent with the discussion you and I had about standardizing column names when new units are involved, and is what I then implemented in the root functions.
- [ ] line 259: `dplyr::group_by_at()` is superceded.
- [ ] line 289: Not needed. Just assign `hbp_agb = hbp_standing_biomass_in_clip_cells` as list argument when making "output.list".
- [ ] line 319: Round "Mg/ha" output to 2 decimal places (i.e., two sig figures) based on precision of input clip strip area data. There are probably a few other places where significant figures ought to be reconsidered in the output given that we are limited by clip area to two significant figures (i.e., 2.0 x 0.1 meters).
- [ ] line 338: The helper text indicates the default for the 'plotPriority' input argument is "5", but the argument definition on line 54 provides "NA". Does something need updating here?

schafnoir commented:
I'll work through the `estimateWoodMass` function next...

schafnoir committed:
Review and code/format standardization for estimateWoodMass function.

schafnoir committed:
More estimateWoodMass review

schafnoir committed:
Continued review of estimateWoodMass

schafnoir committed:
Continued estimateWoodMass review and updates

schafnoir committed:
Reviewed up through Chave AGB estimates

schafnoir committed:
Completed estimateWoodMass review through end of Conti shrub section.

schafnoir committed:
Completed review of estimateWoodMass function

schafnoir commented:
@samsimkin 
I completed review of the `estimateWoodMass` function, with line by line comments below. I re-worked some of the input argument names and the roxygen header material to be as consistent as possible across other functions in the package. Let me know if any of the feedback is unclear! I'll move on to the `estimateMass` function next.

- [x] line 10 (roxygen header): Where can people go to learn more about "plot priority"? Concerned that users will not know how to use this argument correctly. There is no mention of Plot Prioritization in the DPUG, and it is not mentioned in the protocol either. Should we add some limited discussion of the Plot Prioritization list to the DPUG and then reference the DPUG in the function documentation?
- [x] line 29: If only live biomass is returned how is dead biomass estimated at the per plot level? Why not include dead biomass as well and let users filter out if they want only live?  I can see dead mass being needed for some research questions (fuel load?), so seems odd to exclude it from this table.
- [x] line 30: I don't understand how biomass for individuals can be expressed on a per hectare basis. Should units just be "Mg" rather than "Mg/ha"? Additional comments about this below...
- [x] line 121 and lines 182-194: The vst_nonWoody table is defined when list input is provided but not when table input is provided. Appears we need another 'inputNonWoody' argument. Also, the example dataset doesn't have a vst_non-woody table in the list object, which means none of the non-woody code can be tested. Need a test dataset that allows non-woody output and testing. Or remove non-woody code since this function is clearly focused on "wood" per its name?
- [x] line 128: NULL is a placeholder I added. This object should be created from a new table input argument, I believe.
- [x] line 207: Need error handling if plotPriority is not an integer. Also, I think a plotPriority > 30 should trigger an error.
- [x] line 213: The original logic won't work when input is provided via table arguments instead of list argument. I updated to reflect the logic I added on lines 115-130. 
- [ ] line 219-220: How does getting the min/max 'year' across the whole data frame make sense when multiple sites and years may be provided in the input data? Maybe it is just not clear to me yet how these parameters are used later on...
- [ ] lines 278-285: Not clear why duplicates are removed here and a new data frame is created. Maybe used later?
- [x] lines 289-365: Cannot test since test dataset does not contain vst_non-woody data.
- [x] line 294: The original logic to generate the non-woody data frame from the list input will not work if the user does not supply input as a list in the first place. I updated to reflect the logic I added on lines 115-130. 
- [x] line 307: Likely that 6 digits overestimates our true significant figures. I believe height data are to nearest 10 cm, so this will inform sig figs.
- [x] line 315: Why begin to populate the "agb" column at this point? Why not wait until we have all the "other" mass estimated and then calculate "agb" as the sum of all the input columns once they have all been created? Can probably delete this line...
- [x] line 329: 6 digits likely overestimates confidence in mass output.
- [x] line 347: Same as above.
- [x] line 362: Same as above.
- [x] line 398: Why not identify the plantStatus values you do want via plantStatus %in% c("live", "dead"), instead of defining all the ones you don't want?
- [x] line 453: Does it make sense to express the biomass of an individual on a per area basis? Maybe when individuals are summed to the plot, but on strictly an individual basis I am not sure it makes sense to report this as a final output...
- [x] line 483: The original logic won't work when input is provided via table arguments instead of list argument. I updated to reflect the logic I added on lines 115-130. 
- [x] line 505: The original logic won't work when input is provided via table arguments instead of list argument. I updated to reflect the logic I added on lines 115-130. 
- [x] line 608: What about just using a left_join with the 'taxon_fields' and 'plant_char' data frames in one step, rather than 'merge' followed by 'filter'? Recognize this is not hugely important...
- [x] line 870: I am not convinced the assumption to estimate stemDiameter from basalStemDiameter makes sense when stemDiameter is absent. Just looking at the test dataset, the only two individuals to which this approach would be applied are growthForm == "sapling", which by definition have DBH < 1 cm and so estimating DBH for these individuals makes marginal sense since most allometries are not generated from individuals this small. Are there any other cases where stemDiameter would be missing and we WANT to estimate it like this? Nothing comes to mind but I could be missing something...
- [x] line 876: We never measure stemDiameters < 1 cm so I feel like values < 1 cm should simply be excluded. Or we need to implement an approach for these smaller individuals that relies on the data we DO collect.
- [x] line 883: I feel like Fulcrum should prevent entering stemDiameter < 1 cm for the growthForms in question so these sorts of issues in older data probably mean a data entry error and the individual should be excluded (filtered out) rather than have DBH set to 1 cm. Generally, I would rather exclude data than make up something.
- [x] line 890: I think it is equally likely that a data entry error was made for either stemDiameter or basalStemDiameter and we really don't know what the correct value is. I would prefer to filter out these individuals or keep the entered stemDiameter value since we don't 100% know it is wrong - i.e., basalStemDiameter could be wrong.
- [x] line 900: I don't think we can make this assumption about liana outlier diameters without actually looking at distributions of liana stemDiameters to inform a cutoff. I feel like making up a diameter of 1 cm is not any better than accepting what looks like a large value. However, it is a known (and somewhat rare) issue for the measurementHeight to be entered into the stemDiameter field so I am good with taking any stemDiameter value of 130 (for any growthForm) and swapping it with the value in the measurementHeight field.
- [x] line 986: Do I understand correctly that this logic overwrites the AGB for tropical species when "height" is present? Might be able to accomplish this AND the previous step more elegantly with a "case_when()" call.
- [x] line 1011: Why the 'suppressWarnings' call? What are the warnings being generated? I saw no warnings when using test dataset.
- [x] line 1032: I assume this logic needs to be updated to properly handle multi-stem individuals - i.e., calculate round diameter equivalent for all qualifying stems. Also, similar comment about suppressWarnings as above - not seeing any warnings with test dataset. Needed?
- [x] line 1052: Same 'suppressWarnings' comment as above.
- [x] line 1072: Same 'suppressWarnings' comment as above.
- [x] General comment: Having looked at all the shrub biomass code now, it strikes me that a hefty 'case_when()' function call could take care of all the different cases and equations, and then a similar approach could be used to assign the reference source. Not sure if you use 'case_when()', but I love it! Obviously low priority to re-code since current approach works just fine.
- [x] lines 1097-1131: MEPO5 biomass estimation: Test dataset does not allow verification of code output. Maybe not a problem for this limited case, but would be nice to know that the output looks like it is in the right ballpark.
- [x] lines 1136-1150: RHDA biomass estimation: Same comment as for MEPO5.
- [x] lines 1154-1170: ARTR2 biomass estimation: Same comment as for MEPO5. Also, why is 'suppressWarnings()' used here but not for other taxa?
- [x] lines 1174- 1189: What does it mean that the allometric equation for LATR2 was estimated from a figure in the manuscript? Is this reproducible? Also, same comment as for MEPO5 regarding test dataset.
- [x] lines 1194-1217: Same test dataset comment as for MEPO5.
- [x] lines 1221-1234: Same test dataset comment as for MEPO5.
- [x] lines 1238-1251: Same test dataset comment as for MEPO5.
- [x] line 1262: Not sure it is good to assign character "NA", want NA (logical) instead. I updated this code to assign NA (logical) rather than character.
- [x] line 1274: Given the number of different allometries that could be applied to an individual with stemDiameter < 10 cm, I am not convinced it is worth making the assumption that the individual is a small tree if reported growthForm == "unknown".
- [x] line 1342: I think 6 decimal places overestimates our real confidence; likely that two significant figures is the very most we could be confident.
- [x] line 1375: I am not convinced it is sensible to report biomass for individuals on a per hectare basis (i.e., I've never seen this). Individuals by their nature are points in space and have no inherent sampling area associated with them. I understand the "Mg/ha" unit is needed for plot- and site-level table outputs, but I think we should just deliver "Mg" level AGB for the individual table at the very end.
- [ ] line 1419: How are Sampling Impractical values handled in the perplot data when it comes to identifying where real zeroes ought to be assigned? Possible I missed this detail earlier on...
- [x] line 1463: Why is a "__" used in the new column names? I recall discussing with you that when units are used as part of a column name, they should be a suffix like "_Mgha". This plot-level column output seems like something new compared to what we discussed.
- [x] lines 1543-1551: Why no dead mass in the output? Seems like this might be wanted for some research questions (e.g., fire-related fuel loading, etc)?
- [x] lines 1548 and 1550: I think we are limited to 3 sig figs for the mean (1 decimal place, based on our confidence in the plot areas) and two sig figs for the SD (also one decimal place).
- [x] line 1556: Not clear why clean-up is done on 'priority_plots' data frame since it is not used for anything else after this operation and is not delivered to user as part of 'output.list'

schafnoir commented:
@samsimkin - any commits I make moving forward will be for the `estimateMass` function so you can make updates and commits to the `estimateWoodMass` function and we shouldn't create any conflicts.

samsimkin committed:
scaleHerbMass updates following CM review

samsimkin committed:
estimateWoodMass updates

samsimkin committed:
mass functions updates

schafnoir committed:
Reviewing estimateMass roxygen header

schafnoir committed:
Review of documentation complete for estimateMass function

samsimkin committed:
more comprehensive test datasets and productivity function updates

schafnoir committed:
Review of estimateMass() complete

schafnoir commented:
@samsimkin 
I completed review of the `estimateMass` function, with general and line-by-line comments below. Let me know if you have any questions!

**General comments:**
Test dataset: Dataset lacks vst_non-woody data. Brought this up in context of estimateWoodMass() function, but relevant again here (though I think you addressed this already?). Also, there is no BBC dataset for SJER to test this part of the function code. Suggest creating BbcDat.rds and if there is no data for 2021 or 2022 (since BBC is a 5-year protocol), may need to make something artificial - i.e., use whatever year is most recent for SJER and then edit the year to match your existing HBP and VST test data.

Test file: Missing - there is no 'test_estimateMass.R' file.

Roxygen header: Recommend standardizing structure of input table arguments across the data products. Suggested options:

inputVstDataList
inputVstAppInd
inputVstMapTag
inputVstNonWoody
inputVstPerPlot
inputHbpDataList
inputHbpPerBout
inputHbpMass
inputBbcDataList
inputBbcPerCore
inputBbcMass
inputBbcDilution
includeBbcDilution
includeBbcFragment

**Line by line feedback:**
- [ ] line 39: @param plotType - text indicates that "tower" is an option but a complete list of viable options is not provided.
- [ ] line 41: Suggest changing 'plotPriority' to 'towerPlotsSampled' and changing data type from "numeric" input, which can include real numbers, to "character" to constrain people making silly errors. I suggest two options: "annual" and "all", and avoid introducing plot priorities to end-users. May want to update argument in 'estimateWoodMass()' and 'scaleHerbMass()' functions to be the same. I know we discussed this already but just iterating a bit here...
- [ ] line 43: A complete list of viable growthForm options is needed. Don't want to make people guess what the function will accept. Suggest using 3-letter codes and defining them here. Also, what happens if people specify Hbp growthForms but do not include "Hbp" in the dataProducts argument? Probably need additional error handling.
- [ ] line 68: Not sure I understand the default for the 'dataProducts' argument. Why assume that people are default interested in above-ground mass? Maybe just leave it undefined but required?
- [ ] lines 69-85: Likely need to update based on Roxygen header feedback above.
- [ ] line 90: Assuming the user provides input for the 'dataProducts' argument, first need to check that value provided is one of c("Bbc", "Hbp", "Vst") and provide an informative error message if not.
- [ ] line 92: Not clear why the 'inputDataList' object is created from the 'inputDataListVst' object when the latter is then used as an input for the 'estimateWoodMass()' function. What is the logic here? 
- [ ] line 94: Better to use 'message()' than 'print()' according to package dev guidelines I have read. I updated.
- [ ] line 113: Same comment as line 92.
- [ ] line 115: Same comment as line 94.
- [ ] line 129-137: No SJER BBC data in test datasets to verify BBC code works as intended.
- [ ] line 131: Same comment as line 92.
- [ ] line 133: Same comment as line 94.
- [ ] line 149: Same comment as line 94.
- [ ] lines 160-166: Seems like several permutations are missing - i.e., BBC + HBP, and BBC + VST. Also, I think new "mass_Mgha" column can only have 3 significant figures?
- [ ] lines 168-227: There were missing permutations here too, so I re-worked the logic to construct the output list using three simple conditionals. Prior code is commented out so you can see if what I've done is all right. If so, commented code can be deleted.

schafnoir commented:
@samsimkin - I'll review `estimateHerbProd()` next...

samsimkin committed:
Update DESCRIPTION to resolve merge conflict

samsimkin committed:
Merge branch 'dev_SMS' of https://github.com/NEONScience/neonPlants into dev_SMS

samsimkin committed:
mass function updates including new plotSubset arg to replace plotType and plotPriority

samsimkin committed:
update woodMass growthForm filter

samsimkin commented:
@schafnoir Implemented requested feedback, including the implementation of new plotSubset argument we had talked about.

schafnoir committed:
Review of 'estimateHerbProd()' function

schafnoir committed:
Additional review of scaleHerbMass() function

schafnoir commented:
@samsimkin - As I have been going over `estimateHerbProd()`, I noticed in `scaleHerbMass()` that there are options for the 'plotSubset' argument that don't make sense - specifically, "towerAnnualSubset" does not make sense for HBP. In this light, I updated the Roxygen header and lines 141-144, but I think you'll want to re-work any logic that depends on "towerAnnualSubset" after this point.

schafnoir committed:
HBP productivity calculation check

schafnoir commented:
@samsimkin - a few observations and issues discovered during review of `estimateHerbProd()`:

General: Use of SJER dataset is interesting, as `dplyr::last()` approach with bout number isn't going to give the same results for this site as it will for others because sequence of bouts for a given growing season spans two calendar years. This function will need a custom approach to deal with SJER.

- [x] lines 15 and 17: Suggest replacing 'plotType' and 'plotPriority' arguments with the single  'plotSubset' argument now employed for the estimateMass() function. 
- [x] lines 25, 26: Suggest replacing with a call to the scaleHerbMass() function, to illustrate clearly how the output from that function is required.
- [x] lines 31-34: Update with new 'plotSubset' argument.
- [x] lines 41-42: Replace with new 'plotSubset' argument.
- [x] lines 51-52: Deleted. Redundant with call to list2env() immediately above.
- [x] lines 64-75: Rework to check validity of new 'plotSubset' argument.
- [x] line 136: Re-worked to simplify and use single 'stop()' message per CRAN guidelines.
- [x] line 143: Replaced 'print()' with 'message()' per CRAN guidelines.
- [x] lines 149-152: Delete?
- [x] line 173: Delete?
- [x] lines 189-216: Unfortunately, it is not consistent with the Science Design to calculate 'consumption' at the plot level. Feedback from prior TWG members (Alan Knapp) was that our randomly selected "pairs" of clip strips with and without exclosure in the same plot are not close enough to each other to warrant the approach coded here. As such, it is necessary to calculate average biomass across all exclosure clip strips within a bout, then subtract average biomass from ambient clip strips within that same bout. A plot level approach cannot be justified given the sampling design. In "tests" folder, file "hbpProdCheck.R" supports that Science Design approach and plot-level approach do not give the same results (though they are close, and the discrepancy might be due solely to the inability to calculate plot-level consumption when "paired" exclosure == "Y" data are missing - i.e., due to exclosure disturbance by cattle). In any event, having to filter out good "ambient" data for a given subplotID x eventID when the "exclosure" data are missing for that subplotID x eventID combo seems like an unfortunate side-effect of the plot-level approach...

schafnoir committed:
Merge branch 'dev_SMS' into dev_CM

schafnoir committed:
Merge pull request #21 from NEONScience/dev_CM

Merge 'dev' to 'dev_SMS' to simplify check() output

schafnoir committed:
Beginning 'estimateWoodProd()' review

schafnoir committed:
Testing and review of estimateWoodProd()

schafnoir committed:
Continued review of estimateWoodProd() function

schafnoir committed:
Continued review of estimateWoodProd()

schafnoir committed:
Continued review of estimateWoodProd function

schafnoir committed:
Git config and RStudio issues

schafnoir committed:
Git RStudio file struggles

schafnoir committed:
More file struggles

schafnoir committed:
File struggles

schafnoir committed:
File struggles

schafnoir committed:
Strange issues

schafnoir committed:
Continued review of estimateWoodProd() function

schafnoir committed:
Completed review of estimateWoodProd() function

schafnoir committed:
Git ignore updates

samsimkin committed:
estimateHerbProd removes grazed plot records from plot table

In addition, updated test scripts to have inputDataList in single quote to match change in function. Many spaces were auto-removed. Not sure why. Also, .Rd documentation files autoupdated after running devtoola_check based on previous function changes

schafnoir commented:
@samsimkin - feedback for `estimateWoodProd()` function is below...

### High-level recommendations for first iteration of function and package

- [x] Use only "Approach 2" from Clark et al. 2001 to calculate ANPP. This will allow for more easily incorporating NPP from individuals sampled via nested subplot in a future iteration of the function.
- [x] Add "recruitment" to be consistent with Clark et al. 2001
- [x] Remove outlier detection.
- [x] Use "plotSubset" argument and logic to be consistent with other functions. 
- [ ] Consider only calculating ANPP for one interval at a time. May require processing data from only one site at a time as well.
- [x] The test dataset must successfully run and produce sensible output for all lines of code.

### Roxygen header feedback

- [x] Details: Need enough here for a user to make an informed choice about "Approach_1" vs "Approach_2". I would add the formulas from Clark Fig. 2, as well as the implications regarding inclusion/exclusion of smaller growth forms sampled with nested subplots.
- [x] The 'plotType' and 'plotPriority' argument: Replace with 'plotSubset', similar to what we've discussed for other functions.
- [x] The 'outlier' and 'outlierType' arguments: I am leery of including these arguments because I think the default behavior could be undesirable for a user who supplies data that have already been cleaned and who is not paying attention. Perhaps more importantly, there are a number of different ways to detect outliers, and based on my experience with the VST stemDiameter data, I strongly suspect the defaults will lead to removing the biggest trees at a number of sites, and this will have a huge and negative impact on productivity estimates. Instead of building in this functionality, I think it would be better to state in the "Details" that the productivity estimates are very sensitive to outliers, and users should familiarize themselves with the input data (and potential outliers) before using the function.
- [x] Examples, lines 36-37: Replace with call to 'estimateWoodMass()' function because these lines cannot be run by an end-user.

### Line-by-line function code feedback

- [x] Lines 45-49: Update based on comments above about plotPriority and outliers.
- [x] Lines 57-61: Update function arguments based on Roxygen comments above.
- [x] Lines 73-75 are redundant with line 72 --> delete lines 73-75.
- [x] Lines 86-95: Re-work once standardized "plotSubset" argument is implemented.
- [x] Lines 99-101: Check is on "calcMethod" argument, but message is related to "outlierType".
- [x] Lines 104-106: Check is on "outlierType" but message is related to "calcMethod". Probably want to delete these lines anyway because I think outlier detection is problematic.
- [x] Lines 186-190: CRAN guidelines are to embed output into 'stop()' call, rather than precede stop() with two calls to print(). Updated code accordingly.
- [x] Line 199: The default 'calcMethod' is "Approach_1", but calculations begin with "Approach_2", and a message is printed right off the bat talking about "Approach 2" which would show even if calcMethod == "Approach_1". Add a conditional 'if (calcMethod == "Approach_2") {code...}' here? And why not begin with code for "Approach_1" if that is the default? As coded, "approach_2" always runs, regardless of calcMethod input argument.
- [x] Lines 213-220: It is not clear to me why "Dead_or_Lost_Mgha" is created via summarise() then immediately removed. Similarly, why make "Live_Mgha" via summarise() and then immediately rename? Why not just make only what you need via summarise() and then follow up with a call to arrange(.data$year)? 
- [ ] Line 223: I assume this code and the following lines all work with > 2 years of input?
- [x] Line 226-252: The logic and outputs from these lines seem inconsistent with Clark's "Approach 2" as I understand it. In the intermediate 'vst_increment' dataframe, increment is calculated at the scale of the taxonID, which results in a number of NAs for increment when a taxonID is not present in one of the two years. If we assume a stand is roughly equivalent to a plot, Clark Fig 2 indicates to me that all the mass of individuals alive at t2 is summed, and then all the mass of live individuals at t1 is subtracted, and this difference is the "stand" (i.e., plot) level increment; there is no mention of some masses being excluded from the increment calculation based on taxonID. I am confused... why aren't we just using data from the 'vst_agb_kg' data frame as is done for the mortality part of the calculation?
- [x] Lines 244-252: In the 'vst_increment_long' dataframe, a number of the rows are NA for "Mgha_increment", and I don't understand how this is consistent with Clark "Approach_2".
- [x] Line 257: Note that this `message()` will display even when user selects "Approach_1".
- [x] Line 262-267: The 'vst_agb_kg' calculations embedded in the "Approach2" section appear to also be needed for the "Approach1" section (lines 541-552). May need to re-structure the code given that I think a conditional is needed to wrap around the entire "Approach2" section.
- [x] Lines 281-286: After running these lines, there remain a number of individualIDs that have > 1 row with the same "year" and different values for "plantStatus2". I assume this is due to multi-bole 'smt' with different 'plantStatus' for each bole output from 'estimateWoodMass()' function, correct?
- [x] Lines 300-306: Using the test dataset, line 301 doesn't seem to do anything to the commas I see in either of the "status_" columns. As such, lines 302-306 also do not seem to do anything. However, if the code worked, I think the assumption being made is if any bole of a multi-bole 'smt' is alive, the entire individual is considered alive, correct? Since individual boles are not tracked, this seems like a reasonable assumption to make. However, it appears that the logic in these lines is hard-coded based on the order of individual boles' status values seen in the data --> what happens when something like '("Live", "Dead_or_Lost", "Live")' is encountered? Or an 'smt' with > 3 boles? Maybe it would be better to use a 'grepl' call and look for "dead"? Regardless, I think this logic isn't working and may be prone to breaking.
- [x] Lines 309-320: Using the test dataset, every row in the new "transitionType_" column is NA, even for individuals like 'NEON.PLA.D02.BLAN.05237' that should have a value of "mortality" if I am understanding the logic. Moreover, a very large number of individualIDs are "Dead_or_Lost" in 2022 and "NULL" in 2021 --> one way this can arise is if an individual is really dead in 2022 but wasn't sampled in 2021 (but was in 2019 - e.g., 'NEON.PLA.D03.JERC.02229'). There are also a very large number of individuals that are "Live" in 2021 and "NULL" in 2022. As it happens, using BLAN_032 as an example, the nested sub area changed from 2021 to 2022 from 25m2 to 10m2, so a number of individuals had their tags removed and were set to plantStatus = "no longer qualifies". Aside from the obvious problems of the code not working as intended, given the complexity of our design, how can we really be confident about estimating mortality for these smaller individuals? I don't know that we can legitimately treat all growthForms the same when it comes to estimating mortality...
- [x] Lines 331-342: Cannot determine whether code is generating expected outcomes because no rows in 'transitions' dataframe have "mortality" in 'transitionsType_' column.
- [x] Line 344-351: Cannot determine whether "0" for all rows in "Mgha_mortality" column are legit due to upstream issues described above, but this result does not seem reasonable to me. Also, is "year" incremented here because mortality mass in a given year corresponds to mass in the mass in the previously sampled year, but we want to associate this mass with the productivity for the given year?
- [x] Line 386-393: Hard to tell what this logic is doing because outcome with test dataset does not show any changes before/after running these lines.
- [x] Line 410-416: Test dataset does not have records with missing productivity, but how would it be possible to have plots with no productivity at all, assuming there data for two timepoints in the input dataset?
- [x] Line 419-427: Test dataset has 'vst_agb_zeros' with zero rows so cannot evaluate what intent of this logic is.
- [x] Line 442-450: Need to update to use new "plotSubset" argument and logic.
- [x] General question for Approach 2: Where is assessment of in-growth? In Clark Approach 2, authors account for contribution of recruits to productivity by multiplying the number of newly qualifying individuals by the average biomass of trees at the minimum measured diameter. 
- [x] Line 488: Approach1 --> Much of the code related to calculating increment for the earliest year cannot be checked with test dataset. That is, for year "1", all increment data are NA.
- [x] Line 492-493: The 'length()' function calls appear to create unnecessary console output. Can these calls to length() be removed?
- [x] Line 612-625: I think it is key to inform users that the NPP calculations are very sensitive to outliers. However, this filtering is done using normal distribution assumptions, and the code is agnostic with respect to the shape of the distribution associated with the input data. I don't think filtering by outliers should be done in this function since we don't know whether these "outliers" are statistically valid to remove. Even if we were to keep the outlier detection, why calculate outliers for each year - i.e., each iteration of the "for()" loop? Wouldn't it be more valid to identify outliers using data for ALL years after the fact?
- [x] Line 882-901: These lines generate an error with the test dataset, I believe either because all input data are "NA" or because the "Mghayr" column is missing from the "increment" data frame. It seems like "Approach1" cannot operate independently from "Approach2"? Without this code functioning, can't really test whether 'for()' loop is generating sensible results with the test dataset, and cannot test remainder of code.
- [x] Line 985-986: Why only use increment1-3 data for filtering after going to the trouble of calculating increment1-7?
- [x] Line 988: Call to 'length()' will create unnecessary Console output.
- [x] Line 996: Recruitment missing; ANPP output not comparable to Clark "Approach1".
- [x] Line 1058-1068: Logic needs reworking to use new "plotSubset" argument.

schafnoir committed:
Stupid Rproj files

samsimkin committed:
estimateHerbProd updates

samsimkin committed:
estimateWoodProd some updates

samsimkin committed:
estimateWoodProd individ approach removed

samsimkin committed:
estimateWoodProd updates including adding recruitment

At a high level I have included recruitment logic and included that in the woodANPP_Mghayr, expanded the test dataset to include instances of mortality and confirmed that they are producing the desired results, and simplified the code to focus on approach 2 (stand-level). The current script takes as input trees (both single and multi bole trees) from exactly two years, and assumes that the plotSubset is towerAnnualSubset, and for simplicity in v1 I propose that we make these required criteria. Even with just trees, the recruitment logic becomes more complicated when there are 3 or more years worth of data, and as we had previously discussed we would need to have 6 or more years of data for plots sampled only every 5 years. Below I have annotated some of the items in your Git checklist. For the recruitment component you had suggested simply estimating the average biomass of an individual at the minimum diameter threshold. I've done that, at least for v1, applying Chojnacky allometry ID H7 that is essentially a catch-all "other". There may be a better generic allometric equation that could be applied to a generic stem of diameter 10 cm. Alternatively, taxa-specific equations could be applied but that would involve a lot of duplication of effort from the estimateWoodMass function. Since we are talking about small (10 cm diameter) individuals this might not matter.

Line 226-252: The logic and outputs from these lines seem inconsistent with Clark's "Approach 2" as I understand it. In the intermediate 'vst_increment' dataframe, increment is calculated at the scale of the taxonID, which results in a number of NAs for increment when a taxonID is not present in one of the two years. If we assume a stand is roughly equivalent to a plot, Clark Fig 2 indicates to me that all the mass of individuals alive at t2 is summed, and then all the mass of live individuals at t1 is subtracted, and this difference is the "stand" (i.e., plot) level increment; there is no mention of some masses being excluded from the increment calculation based on taxonID. I am confused... why aren't we just using data from the 'vst_agb_kg' data frame as is done for the mortality part of the calculation?
 - the plot-level biomass from the estimateWoodMass should be appropriate for the stand-level increment here (and the summary steps already made). Normally the taxonIDs would sum nicely and we would end up at the same place. Here there is an artifact of an eventID which according to perplotperyear was sampled, yet had no records in the filtered output df, and was subsequently populated with zeros, but not the nlcdClass or taxonID so those fields were NA. For simplicity I have removed the nlcdClass and taxonID grouping variables

Lines 244-252: In the 'vst_increment_long' dataframe, a number of the rows are NA for "Mgha_increment", and I don't understand how this is consistent with Clark "Approach_2".
  - this is an artifact of the issue discussed above. For simplicity and to avoid this edge artifact both the nlcdClass and taxonID grouping variables have been remove

Lines 281-286: After running these lines, there remain a number of individualIDs that have > 1 row with the same "year" and different values for "plantStatus2". I assume this is due to multi-bole 'smt' with different 'plantStatus' for each bole output from 'estimateWoodMass()' function, correct?
 - I am not able to reproduce this (with growthForm = "tree"), so I suspect you are right

Lines 300-306: Using the test dataset, line 301 doesn't seem to do anything to the commas I see in either of the "status_" columns. As such, lines 302-306 also do not seem to do anything. However, if the code worked, I think the assumption being made is if any bole of a multi-bole 'smt' is alive, the entire individual is considered alive, correct? Since individual boles are not tracked, this seems like a reasonable assumption to make. However, it appears that the logic in these lines is hard-coded based on the order of individual boles' status values seen in the data --> what happens when something like '("Live", "Dead_or_Lost", "Live")' is encountered? Or an 'smt' with > 3 boles? Maybe it would be better to use a 'grepl' call and look for "dead"? Regardless, I think this logic isn't working and may be prone to breaking.
 - Thanks for catching this. This must have worked originally, but as you note it was not currently working with the current test dataset. This is now corrected.

 Lines 309-320: Using the test dataset, every row in the new "transitionType_" column is NA, even for individuals like 'NEON.PLA.D02.BLAN.05237' that should have a value of "mortality" if I am understanding the logic. Moreover, a very large number of individualIDs are "Dead_or_Lost" in 2022 and "NULL" in 2021 --> one way this can arise is if an individual is really dead in 2022 but wasn't sampled in 2021 (but was in 2019 - e.g., 'NEON.PLA.D03.JERC.02229'). There are also a very large number of individuals that are "Live" in 2021 and "NULL" in 2022. As it happens, using BLAN_032 as an example, the nested sub area changed from 2021 to 2022 from 25m2 to 10m2, so a number of individuals had their tags removed and were set to plantStatus = "no longer qualifies". Aside from the obvious problems of the code not working as intended, given the complexity of our design, how can we really be confident about estimating mortality for these smaller individuals? I don't know that we can legitimately treat all growthForms the same when it comes to estimating mortality...
 - individuals that experienced mortality (transition from Live to Dead) are now again identified as such. Code now simplified to demand only growthForm of "tree" (single or multi bole tree).

Lines 331-342: Cannot determine whether code is generating expected outcomes because no rows in 'transitions' dataframe have "mortality" in 'transitionsType_' column.
  - upstream issue addressed and a plot with a tree experiencing mortality is now included, so non-zero mortality now produced when expected

 Line 344-351: Cannot determine whether "0" for all rows in "Mgha_mortality" column are legit due to upstream issues described above, but this result does not seem reasonable to me. Also, is "year" incremented here because mortality mass in a given year corresponds to mass in the mass in the previously sampled year, but we want to associate this mass with the productivity for the given year?
  - as for last comment, upstream issue addressed and a plot with a tree experiencing mortality is now included, so non-zero mortality now produced when expected. And yes to the question.

Line 386-393: Hard to tell what this logic is doing because outcome with test dataset does not show any changes before/after running these lines.
  - as for previous 2 comments, upstream issue addressed

Simplified version of script now has code for individual-level approach 1 removed, so last 8 comments not addressed

samsimkin committed:
neonPlants estimateWoodProd supporting towerALL now as well

samsimkin committed:
estimateWoodProd correction

schafnoir committed:
Updates to finalize 'scaleHerbMass' function

schafnoir committed:
Finalized scaleHerbMass function, updated associated tests to be complete with output.

schafnoir committed:
Updated scaleHerbMass man page

schafnoir committed:
scaleHerbMass to dev_CM branch

schafnoir committed:
Add scaleHerbMass to namespace

schafnoir committed:
Add priority_plots to 'data'

schafnoir committed:
Issues with "priority_plots" dataset

schafnoir committed:
Finalized scaleHerbMass function, documentation, and tests.

schafnoir committed:
Merge branch 'dev_SMS' into dev_CM

schafnoir committed:
Merge pull request #22 from NEONScience/dev_CM

Dev cm

schafnoir committed:
Rproj difficulties

schafnoir committed:
Required commit to switch branches, confusing

schafnoir committed:
Test file formattign for estimateHerbProd

schafnoir committed:
Rproj files

samsimkin committed:
redo fix

schafnoir commented:
Closing - want to create PRs for function-specific branches.
